File: golang/defaultDataWriter.go
package abi

import (
	"encoding/hex"
	"errors"
	"strings"
)

type defaultDataWriter struct {
	parts [][]byte
}

// NewDefaultDataWriter creates a new defaultDataWriter.
// A newly-created writer has no parts.
// Parts are created by calling GotoNextPart().
func NewDefaultDataWriter() *defaultDataWriter {
	return &defaultDataWriter{
		parts: [][]byte{},
	}
}

func (d *defaultDataWriter) Write(data []byte) (int, error) {
	if len(d.parts) == 0 {
		return 0, errors.New("cannot write, since there is no part to write to")
	}

	partIndex := len(d.parts) - 1
	d.parts[partIndex] = append(d.parts[partIndex], data...)
	return len(data), nil
}

func (d *defaultDataWriter) GotoNextPart() {
	d.parts = append(d.parts, []byte{})
}

func (d *defaultDataWriter) GetParts() [][]byte {
	return d.parts
}

func (d *defaultDataWriter) String() string {
	partsHex := make([]string, len(d.parts))

	for i, part := range d.parts {
		partsHex[i] = hex.EncodeToString(part)
	}

	return strings.Join(partsHex, partsSeparator)
}

File: python/multiversx_sdk_abi/default_data_writer.py

from typing import List

from python.multiversx_sdk_abi.constants import PARTS_SEPARATOR
from python.multiversx_sdk_abi.errors import \
    ErrWriterCannotWriteSinceThereIsNoPart


class DefaultDataWriter:
    def __init__(self):
        """
        Creates a new DefaultDataWriter.
        A newly-created writer has no parts.
        Parts are created by calling goto_next_part().
        """
        self.parts: List[bytes] = []

    def write(self, data: bytes):
        if len(self.parts) == 0:
            raise ErrWriterCannotWriteSinceThereIsNoPart()

        part_index = len(self.parts) - 1
        self.parts[part_index] += data

    def goto_next_part(self):
        self.parts.append(b"")

    def get_parts(self) -> List[bytes]:
        return self.parts

    def __str__(self):
        parts_hex = [part.hex() for part in self.parts]
        return PARTS_SEPARATOR.join(parts_hex)
